### ISSUE TITLE
Improve bundle size

### ISSUE THREAD

**Issue by DavidRouyer** [neondatabase/serverless#48]
## Steps to reproduce

I'm building a project with Next.js, tRPC, Drizzle and Neon as a Postgres database (you can find the source code here: https://github.com/DavidRouyer/customer-service)

The library is used in the following way:

```ts
import { neon } from '@neondatabase/serverless';
import { drizzle } from 'drizzle-orm/neon-http';

import * as auth from './schema/auth';
import * as contact from './schema/contact';
import * as message from './schema/message';
import * as ticket from './schema/ticket';

export const schema = { ...auth, ...contact, ...message, ...ticket };

export { pgTable as tableCreator } from './schema/_table';

export * from 'drizzle-orm';

export const db = drizzle(neon(process.env.DATABASE_URL!), { schema });
```

I'm having an issue when deploying my application on Vercel, I'm getting the following message: `The Edge Function "index" size is 1.04 MB and your plan size limit is 1 MB.`

I've installed `@next/bundle-analyzer` to investigate the problem with the bundle size on the edge part:

![image](https://github.com/neondatabase/serverless/assets/1556750/74451059-df2a-4781-b3ac-3c6af94693b9)

As you can see, `@neondatabase/serverless` is bundled multiple times and represents a large part of my bundle:

![image](https://github.com/neondatabase/serverless/assets/1556750/96faff17-d17c-474d-9ef3-1896dcc3de83)

## Expected result
Have a smaller bundle size (or maybe a specific entry for edge only?)

## Actual result
The bundle size is too big

## Environment
Vercel, Edge Runtime

## Logs, links
-

**Comment by DavidRouyer** [comment]
If I remove test functions from `index.ts`, the file goes from `dist/serverless.mjs  185.5kb` to `dist/serverless.mjs  137.4kb` when building

**Comment by jawj** [comment]
Hi @DavidRouyer. Thanks for filing this issue.

On your latest point, `dist/serverless.mjs` is not part of the npm package. The main bundled code in the npm package is in `dist/npm/index.js`, which is similar in size (at 140KB) to what you got by removing tests.

A key issue appears to be that at least 3 copies of our package are being included by the Next.js bundler. Do you think that's normal/unavoidable?

If you're only importing `{ neon }` from the package, only a small fraction of the codebase should ever be used, so it's a shame that tree-shaking isn't eliminating the rest. I'll see if there's anything we can do about this.

**Comment by DavidRouyer** [comment]
Thank you for your answer!

> On your latest point, `dist/serverless.mjs` is not part of the npm package. The main bundled code in the npm package is in `dist/npm/index.js`, which is similar in size (at 140KB) to what you got by removing tests.

Yeah I removed my comment because I was building the project with `npm run build` and not `npm run export`

> A key issue appears to be that at least 3 copies of our package are being included by the Next.js bundler. Do you think that's normal/unavoidable?

> If you're only importing `{ neon }` from the package, only a small fraction of the codebase should ever be used, so it's a shame that tree-shaking isn't eliminating the rest. I'll see if there's anything we can do about this.

I don't know, I'm fairly new to to Next.js and everything is obfuscated, I'm just doing a standard `next build`.
I'm looking for a solution to include the package only once, or tree shaking it.

**Comment by DavidRouyer** [comment]
I replaced the `esbuild` script with `Vite` (esbuild & rollup), and I got better results out of the box https://github.com/neondatabase/serverless/commit/a3acd190a1af3397240fb9271c72af75304a2323

✓ 114 modules transformed.
`dist/index.mjs  140.58 kB │ gzip: 38.93 kB` (unminified)
`dist/index.umd.js  103.74 kB │ gzip: 32.86 kB`

**Comment by jawj** [comment]
I've looked at this a bit more, and not found any easy way to reduce the bundle size that doesn't have drawbacks.

If we remove methods from `Buffer` that aren't used, buffers returned by the library (e.g. for `bytea` columns) will be lacking expected methods.

And skipping `--keep-names` peppers error messages and stack traces with cryptic variable names.

**Comment by nicksrandall** [comment]
@jawj I was just coming in to report this issue -- glad it is already being discussed. For my specific use-case, I *only* need to use the `neon` export so it would be awesome if there was an entry to this package that only exports that function and nothing else (like `Client` and `Pool`). 

Relying on your user's bundlers to do the magic of tree shaking isn't a great strategy IMO. It's better to provide the individual modules yourself if possible (where you *can* control the bundler).

**Comment by jawj** [comment]
Thanks @nicksrandall. Unfortunately even the `{ neon }` import pulls in a fair bit of `node-postgres`, in order to maintain compatibility by using its pre- and post-processing routines on the data that goes in and out of the database.

But I'll keep the issue open for a bit and see if there's anything we can do here.

**Comment by jawj** [comment]
We've reduced the bundle size (including dependencies) by no longer depending on Node or pg types. https://github.com/neondatabase/serverless/pull/206

Client, Pool and the http function are still all bundled in one package, though.
